modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTask.java

```diff
@@ -295,26 +298,29 @@ private static ElasticsearchException wrapException(Exception ex) {
         return storedContext;
     }
 
-    private class CheckpointHandler implements Reindexer.CheckpointListener {
-        private ReindexTaskIndexState lastState;
-        private Semaphore semaphore = new Semaphore(1);
+    private class ProgressState implements Reindexer.CheckpointListener {
 
-        CheckpointHandler(ReindexTaskIndexState lastState) {
-            this.lastState = lastState;
+        private final Semaphore semaphore = new Semaphore(1);
+        private ReindexTaskState lastState;
+        private boolean isDone = false;
+
+        ProgressState(ReindexTaskState initialState) {
+            this.lastState = initialState;
         }
 
         @Override
         public void onCheckpoint(ScrollableHitSource.Checkpoint checkpoint, BulkByScrollTask.Status status) {
             // todo: need some kind of throttling here, no need to do this all the time.
             // only do one checkpoint at a time, in case checkpointing is too slow.
-            if (semaphore.tryAcquire()) {
-                ReindexTaskIndexState nextState = lastState.withCheckpoint(checkpoint, status);
-                this.lastState = nextState;
+            if (isDone == false && semaphore.tryAcquire()) {
```

Comment:
I think you have to swap the order here so that you only check isDone after having obtained the semaphore. Otherwise you risk seeing `isDone==false`, then obtaining the semaphore after `isDone` was set to true.

```suggestion
            if (semaphore.tryAcquire() && isDone == false) {
```